README.md

[![Driven Network & Pagekit Banner](https://driven.network/storage/github-driven-pagekit.png "Driven Network & Pagekit Banner")](https://driven.network/ "Driven Network, LLC")

Pagekit Manifest
================

Web App Manifest Generator for your [Pagekit](https://pagekit.com/ "Pagekit CMS") project. 
Learn more about [PWA (Progressive Web Apps)](https://developers.google.com/web/progressive-web-apps/ "Learn about Progressive Web Apps") and the [Web App Manifest](https://developer.mozilla.org/en-US/docs/Web/Manifest "Learn about Web App Manifest")

Changelog
---------
[View Log](https://github.com/DrivenNetwork/pagekit-manifest/blob/master/CHANGELOG.md "Pagekit Listings Changelog") of releases and updates.

Copyright
---------
Copyright [driven.](https://driven.network/ "Driven Network, LLC") under the [MIT license](https://github.com/pagekit/pagekit/blob/develop/LICENSE "MIT License").


CHANGELOG.md
# Changelog

## 1.0.3 - January 10, 2017

### Icon Paths Issue
- Fix paths issues with icons not being rendered correctly.

## 1.0.0 - November 28, 2016

### Initial Commit
- Upload of all code and files for version 1.0



js/manifest.js
$(function(){

    new Vue({
        el:'#manifest',

        data:{
            manifest: $data.manifest,
            options:{
                icons: {
                    size: [28, 36],
                    density: [0.75, 1, 1.5, 2, 3, 4],
                    type: ['image/gif', 'image/png', 'image/x-icon','image/webp'],
                    extention: ['gif', 'png', 'ico', 'webp']
                },
                dir:['ltr', 'rtl','auto'],
                display:['fullscreen','standalone','minimal-ui','browser'],
                orientation:['any','natural','landscape','landscape-primary','landscape-secondary','portrait','portrait-primary','portrait-secondary']
            }
        },

        methods: {
            save: function(){
                this.$http.post('admin/manifest/save', {manifest: this.manifest}).then(
                function(){
                    UIkit.notify('Saved');
                }).catch( function(){
                    UIkit.notify('Couldn\'t Save');
                })
            }
        }
    });

});


src/Controller/ManifestController.php
<?php

namespace Pagekit\Manifest\Controller;

use Pagekit\Application as App;


class ManifestController
{

    /**
     * @Access(admin=true)
     */
    public function indexAction()
    {

        $module = App::module('manifest');
        $config = $module->config;

        return [
            '$view' => [
                'title'=>'',
                'name'=>'manifest:views/index.php'
            ],
            'manifest'=> $config['manifest'],
            '$data'=> $config
        ];
    }

    /**
     * @Access(admin=true)
     * @Request({"manifest"}, csrf=true)
     */
    public function saveAction($manifest)
    {
        App::config('manifest')->set('manifest', $manifest);
        return App::response('Success', 200 );
    }

    public function webmanifestAction()
    {

        $module = App::module('manifest');
        $config = $module->config;
        $response = str_replace("storage", "\\/storage", json_encode($config['manifest']));
        return App::response($response, 200, ['Content-Type' => 'application/manifest+json'.'; charset='.'UTF-8']);
    }

}

index.php
<?php

// packages/driven/manifest/index.php

return [

    'name' => 'manifest',

    'type' => 'extension',

    // array of namespaces to autoload from given folders
    'autoload' => [
        'Pagekit\\Manifest\\' => 'src'
    ],

    'settings' => '/admin/manifest',

    'resources'=>[
        'manifest:'=>''
    ],

    'events'=>[
        'boot'=> function ($event, $app){

            $app->on('view.meta', function ($event, $meta) use ($app) {

                $meta([
                    'link:manifest' => [
                        'href' => '/manifest/webmanifest',
                        'rel' => 'manifest',
                        'type' => 'application/manifest+json'
                    ]
                    ]);

            }, 50);
        }

    ],

    // array of routes
    'routes' => [

        // identifier to reference the route from your code
        '@manifest' => [

            // which path this extension should be mounted to
            'path' => '/manifest',

            // which controller to mount
            'controller' => 'Pagekit\\Manifest\\Controller\\ManifestController'
        ]
    ],

    'menu' => [
        'manifest' => [
            'label'=>'Manifest',
            'icon' => 'manifest:icon.svg',
            'url'=>'@manifest'
        ]
    ],

    /**
     * Define theme configuration.
     * This is the theme's default configuration. During run-time, they will be merged with
     * settings the user has set in the backend. The default configuration can therefore
     * be overwritten.
     */
    'config' => [
        'manifest' => [
            'lang' => 'en',
            'name' => '',
            'short_name' => '',
            'description' => '',
            'scope' => '/',
            'start_url' => '/',
            'display' => 'standalone',
            'orientation' => 'any',
            'theme_color' => '#cccccc',
            'background_color' => '#cccccc',
            'icons' => array([
                    'src'=> '',
                    'size'=> '36x36',
                    'type'=> 'image/png',
                    'density' => '0.75'
                    ],[
                    'src'=> '',
                    'size'=> '48x48',
                    'type'=> 'image/png',
                    'density' => '1.0'
                ],[
                    'src'=> '',
                    'size'=> '72x72',
                    'type'=> 'image/png',
                    'density' => '1.5'
                ],[
                    'src'=> '',
                    'size'=> '96x96',
                    'type'=> 'image/png',
                    'density' => '2.0'
                ],[
                    'src'=> '',
                    'size'=> '144x144',
                    'type'=> 'image/png',
                    'density' => '3.0'
                ],[
                    'src'=> '',
                    'size'=> '192x192',
                    'type'=> 'image/png',
                    'density' => '4.0'
                ]
            )
        ]
    ]
];
